subsys/zigbee/osif/zb_nrf_serial.c

```diff
@@ -0,0 +1,489 @@
+			     i++) {
+				char_handler(evt->data.rx.buf[i]);
+			}
+		}
+
+		if (uart_rx_buf) {
+			restart_rx = handle_rx_chunk(evt->data.rx_buf.buf,
+				       evt->data.rx.offset + evt->data.rx.len);
+		}
+		break;
+
+	case UART_RX_BUF_REQUEST:
+		/*
+		 * Sync: The continuous RX is implemented by receiving data
+		 *       through chunks into static memory.
+		 */
+		/* Async: The continuous RX is implemented by passing chunk
+		 *        buffers while the user buffer is unknown and filling
+		 *        the user's buffer afterwards.
+		 */
+		if (uart_rx_buf &&
+		    !blocked_buf_chunk &&
+		    uart_rx_buf_len > ZIGBEE_UART_RX_CHUNK_SIZE * 2) {
+			/*
+			 * Async: The user provided RX buffer and the driver
+			 *        has just received the first chunk of data,
+			 *        which is already copied to user's buffer.
+			 *        The second chunk is being received, so
+			 *        skip two chunks while passing the
+			 *        user's buffer.
+			 */
+			uart_rx_buf_rsp(uart_dev,
+			    &(uart_rx_buf[ZIGBEE_UART_RX_CHUNK_SIZE * 2]),
+			    uart_rx_buf_len - ZIGBEE_UART_RX_CHUNK_SIZE * 2);
+
+			/*
+			 * Async: Remember the free RX chunk and restore it once
+			 *        user's RX buffer will be filled with data.
+			 */
+			blocked_buf_chunk = uart_rx_buf_chunk_released;
+		} else if (uart_rx_buf_chunk_released) {
+			/* Sync: Pass the most recently released buffer. */
+			uart_rx_buf_rsp(uart_dev,
+					uart_rx_buf_chunk_released,
+					ZIGBEE_UART_RX_CHUNK_SIZE);
+		}
+
+		/* The RX chunk was either remembered or used for reception. */
+		uart_rx_buf_chunk_released = NULL;
+		break;
+
+	case UART_RX_BUF_RELEASED:
+		if (uart_rx_buf) {
+			if (uart_rx_buf_offset == uart_rx_buf_len) {
+				uart_rx_buf_len = 0;
+				zigbee_schedule_callback(uart_rx_notify, 0);
+			}
+		}
+
+		/*
+		 * Pass the consumed buffer through uart_rx_buf_chunk_released
+		 * variable to keep RX chunks in a loop.
+		 */
+		if (!uart_rx_buf_chunk_released) {
+			if ((evt->data.rx_buf.buf == uart_rx_buf_chunk[0]) ||
+			    (evt->data.rx_buf.buf == uart_rx_buf_chunk[1])) {
+				uart_rx_buf_chunk_released =
+					evt->data.rx_buf.buf;
+			} else if (blocked_buf_chunk) {
+				uart_rx_buf_chunk_released = blocked_buf_chunk;
+				blocked_buf_chunk = NULL;
+			}
+		}
+		break;
+
+	case UART_RX_DISABLED:
+		if (restart_rx && (!is_sleeping)) {
+			/*
+			 * Async: Reception was stopped due to RX timeout,
+			 *        restart it automatically.
+			 */
+			restart_rx = false;
+			uart_rx_buf_chunk_released = uart_rx_buf_chunk[1];
+			(void)uart_rx_enable(
+				uart_dev,
+				uart_rx_buf_chunk[0],
+				ZIGBEE_UART_RX_CHUNK_SIZE,
+				CONFIG_ZIGBEE_UART_PARTIAL_RX_TIMEOUT);
+		} else {
+			/*
+			 * Reception finished and the next RX operation may
+			 * be started.
+			 */
+			restart_rx = false;
+			k_sem_give(&rx_done_sem);
+		}
+		break;
+
+	case UART_RX_STOPPED:
+		/* An error occurred during reception. Abort RX. */
+		if (rx_data_cb && uart_rx_buf) {
```

Comment:
`rx_data_cb` should be rather checked in `uart_rx_notify()` (or perhaps `zb_osif_serial_recv_data()` should not be allowed without the callback installed first, or is there a use case for such scenario?) instead of here, as `uart_rx_notify()` is actually dismissing the user buffer and releasing `rx_fill_buffer_sem` (or this part of code should be moved from there).

Reply:
Yes, `rx_data_cb` should be checked inside `uart_rx_notify`.
I do not see a scenario, in which the reception is scheduled,while the RX callback is not set. Additional fail-safe check added.
The `uart_rx_notify` not only calls the user callback, but also resets the user buffer state, which is necessary for the correct operation.
The `rx_data_cb` still can be set/changed/reset during reception, so all checks for its value are still inside the code.